src/expression.d

```diff
@@ -2831,7 +2831,22 @@ extern (C++) abstract class Expression : RootObject
             assert(type);
             if (!type.isMutable())
             {
-                error("cannot modify %s expression %s", MODtoChars(type.mod), toChars());
+                bool errFromFunc;
+                for (Dsymbol s = sc.func; s; s = s.toParent2())
+                {
+                    FuncDeclaration ff = s.isFuncDeclaration();
+                    if (!ff)
+                        break;
+                    if (!ff.type.isMutable)
```

Comment:
What if the expression does not use `this`? If if doesn't, then the const-ness of the function is not relevant.